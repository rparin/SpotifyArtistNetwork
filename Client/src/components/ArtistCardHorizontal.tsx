import React from "react";
import { ExternalLink } from "lucide-react";
import Image from "next/image";
import HorizontalList from "./UI/HorizontalList";
import { formatNumber, imageLoader } from "@/lib/utils";

export interface artistCardType extends React.HtmlHTMLAttributes<HTMLElement> {
  name: string;
  img: string;
  alt: string;
  genres: Array<string>;
  followers: string;
  pop: string;
  url: string;
  id: string;
}

const ArtistCardHorizontal = React.forwardRef<HTMLElement, artistCardType>(
  ({ id, name, img, alt, genres, followers, pop, url }, ref) => {
    return (
      <section ref={ref} className="relative w-96 rounded-md">
        <article className="flex items-center rounded-md py-2 gap-2 backdrop-blur-md px-2">
          <Image
            loader={() => {
              return imageLoader(img);
            }}
            className=" h-24 w-24 object-cover rounded-[50%]"
            width={100}
            height={100}
            src={img}
            alt={alt}
          />
          <div className="flex flex-col gap-1 w-64">
            <h2 className="flex text-xl font-medium line-clamp-1 justify-center text-center">
              {name}
            </h2>
            <div className="grid grid-cols-2 gap-1">
              <div className="flex gap-2 bg-slate-200/40 dark:bg-slate-600/60 py-1 justify-center">
                <h3 className="text-xs font-normal">Followers</h3>
                <p className="font-medium text-xs">
                  {formatNumber(parseInt(followers))}
                </p>
              </div>
              <div className="flex gap-2 bg-slate-200/40 dark:bg-slate-600/60 py-1 justify-center">
                <h3 className="text-xs font-normal">Popularity</h3>
                <p className="font-medium text-xs">{pop}</p>
              </div>
            </div>
            {genres.length > 0 && (
              <HorizontalList
                id={id}
                items={genres}
                className="dark:bg-slate-500/80 bg-slate-200/40 text-xs"
              />
            )}

            <a
              href={url}
              target="_blank"
              rel="noopener noreferrer"
              className="flex gap-1 text-xs justify-center items-center text-center w-full rounded-br-md bg-[#1DB954] horizontal-mask">
              Spotify
              <ExternalLink className="w-auto h-[.7rem] mb-2" />
            </a>
          </div>
        </article>
        <Image
          loader={() => {
            return imageLoader(img);
          }}
          className="absolute top-0 left-0 -z-10 h-full w-full opacity-80 dark:opacity-40 rounded-md"
          width={100}
          height={100}
          src={img}
          alt={alt}
        />
      </section>
    );
  }
);

ArtistCardHorizontal.displayName = "ArtistCardHorizontal";

export { ArtistCardHorizontal };
