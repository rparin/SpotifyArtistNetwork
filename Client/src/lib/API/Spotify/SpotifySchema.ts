import { z } from "zod";

export const spotifyCTokenSchema = z
  .string()
  .length(115, { message: "Invalid token length" });
