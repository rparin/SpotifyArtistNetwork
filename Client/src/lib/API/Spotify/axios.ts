import axios from "axios";

const BASE_URL = `${process.env.NEXT_PUBLIC_SERVER_URL}`;

export const spotifyAPI = axios.create({
  baseURL: BASE_URL,
});
