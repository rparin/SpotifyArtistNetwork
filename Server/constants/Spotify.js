const Urls = {
  api: "https://api.spotify.com/v1",
  accounts: "https://accounts.spotify.com",
  host: "http://localhost:8080",
  authRedirect: "http://localhost:3000/mynetwork",
};

function getUser(id) {
  return `${Urls.api}/users/${id}`;
}

function getArtists(id) {
  return `${Urls.api}/artists/${id}`;
}

function getRelatedArtists(id) {
  return `${Urls.api}/artists/${id}/related-artists`;
}

function searchArtist(artist) {
  return `${Urls.api}/search?q=${artist}&type=artist`;
}

function searchQuery(query) {
  return `${Urls.api}/search?${query}`;
}

const HeaderContent = {
  urlencoded: "application/x-www-form-urlencoded",
};

function getBasicHeader(authToken) {
  return {
    headers: {
      Authorization: `Basic ${authToken}`,
      "Content-Type": HeaderContent.urlencoded,
    },
  };
}

function getBearerHeader(tokenType, authToken) {
  return {
    headers: {
      Authorization: `${tokenType} ${authToken}`,
    },
  };
}

const Endpoints = {
  authCallback: `${Urls.host}/api/spotify/authCallback`,
  authRedirect: `${Urls.authRedirect}`,
  auth: `${Urls.accounts}/authorize?`,
  token: `${Urls.accounts}/api/token`,
  myInfo: `${Urls.api}/me`,
  myFollowingArtists: `${Urls.api}/me/following?type=artist&limit=50`,
  searchQuery: searchQuery,
  getArtists: getArtists,
  getRelatedArtists: getRelatedArtists,
  searchArtist: searchArtist,
  getUser: getUser,
};

const Headers = {
  basic: getBasicHeader,
  bearer: getBearerHeader,
};

const Grants = {
  refreshToken: "refresh_token",
  authCode: "authorization_code",
  clientCred: "client_credentials",
};

const Variables = {
  delay: 1000,
  randomBytes: 20,
  userScope: "user-follow-read user-top-read", //user-read-private user-read-email
};

module.exports = {
  Endpoints,
  Grants,
  Headers,
  Variables,
};
