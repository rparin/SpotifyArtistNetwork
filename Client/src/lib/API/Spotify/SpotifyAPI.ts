import { spotifyAPI } from "./axios";
import { useQuery, useInfiniteQuery } from "@tanstack/react-query";

const fetchSearchResults = async (
  artist: string,
  page: string | null | undefined,
  accessToken: string
) => {
  const res = await spotifyAPI.get(
    `/api/spotify/search/${artist}/${page}/${accessToken}`
  );
  return res.data;
};

const getClientToken = async () => {
  const res = await spotifyAPI.get("/api/spotify/cAuthToken");
  return res.data;
};

const getMyFollowingArtists = async (accessToken: string) => {
  const res = await spotifyAPI.get(
    `/api/spotify/myFollowingArtists/${accessToken}`
  );
  return res.data;
};

const fetchArtistNetwork = async (
  id: string,
  depth: string,
  accessToken: string
) => {
  const res = await spotifyAPI.get(
    `/api/spotify/relatedMap/${id}/${depth}/${accessToken}`
  );
  return res.data;
};

export const useSpotifyCToken = (enabled: boolean) => {
  const cToken = useQuery({
    enabled: enabled,
    queryKey: ["cToken"],
    queryFn: getClientToken,
    refetchOnWindowFocus: true,
    refetchInterval: 3000000, //grab new token every 50 minutes
    staleTime: 3000000, //cache token for 50 minutes
  });
  return cToken;
};

export const useGetFollowingArtists = (
  enabled: boolean,
  accessToken: string
) => {
  const cToken = useQuery({
    enabled: enabled,
    queryKey: ["followingArtists"],
    queryFn: () => {
      return getMyFollowingArtists(accessToken);
    },
    staleTime: 3000000, //cache data for 50 minutes
  });
  return cToken;
};

export const useGetNetworkQuery = (
  enabled: boolean,
  artistId: string,
  depth: string,
  accessToken: string
) =>
  useQuery({
    enabled: enabled,
    queryKey: ["network", artistId],
    queryFn: async () => {
      return fetchArtistNetwork(artistId, depth, accessToken);
    },
    staleTime: 1800000, //cache network data for 30 minutes
  });

export const useGetSearchQuery = (
  enabled: boolean,
  query: string,
  accessToken: string
) => {
  const parseNextPage = (nextPage: string) => {
    const searchParams = new URLSearchParams(nextPage);
    return `query=${query}&type=artist&offset=${searchParams.get(
      "offset"
    )}&limit=${searchParams.get("limit")}`;
  };
  return useInfiniteQuery({
    enabled: enabled,
    queryKey: ["query", query],
    queryFn: (params: any) => {
      return fetchSearchResults(query, params.pageParam, accessToken);
    },
    initialPageParam: undefined,
    getNextPageParam: (lastPage: any) => {
      if (lastPage.artists.next == null) return null;
      return parseNextPage(lastPage.artists.next);
    },
    staleTime: 600000, //cache search for 10 minutes
  });
};
