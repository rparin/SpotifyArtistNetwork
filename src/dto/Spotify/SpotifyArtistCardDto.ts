import { SPOTIFY_ARTIST_ID_LENGTH } from "@/constants/SpotifyConstants";
import { z } from "zod";

export const spotifyArtistCardDto = z.object({
  id: z.string().length(SPOTIFY_ARTIST_ID_LENGTH),
  name: z.string(),
  img: z.string().url().nullable(),
  followers: z.number(),
  pop: z.number(),
  genres: z.array(z.string()),
  url: z.string().url(),
  alt: z.string(),
});

export type SpotifyArtistCard = z.infer<typeof spotifyArtistCardDto>;
