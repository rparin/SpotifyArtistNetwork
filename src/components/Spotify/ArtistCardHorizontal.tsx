import React from "react";
import Image from "next/image";
import GenreTags from "./GenreTags";
import abbreviateNumber from "@/utils/abbreviateNumber";
import { ExternalLink } from "lucide-react";
import { NO_IMAGE } from "@/constants/SpotifyConstants";
import { SpotifyArtistCard } from "@/dto/Spotify/SpotifyArtistCardDto";

const ArtistCardHorizontal = React.forwardRef<HTMLElement, SpotifyArtistCard>(
  ({ id, name, img, genres, followers, pop, url, alt }, ref) => {
    return (
      <section ref={ref} className="relative w-[22rem] rounded-md md:w-96">
        <div className="flex items-center gap-2 rounded-md px-2 py-2 backdrop-blur-md">
          <Image
            className="h-20 w-20 rounded-[50%] object-cover md:h-24 md:w-24"
            width={100}
            height={100}
            src={img ? img : NO_IMAGE}
            alt={alt}
          />
          <div className="flex w-60 flex-col gap-1 md:w-64">
            <h2 className="line-clamp-1 flex justify-center text-center text-sm font-medium leading-tight md:text-base md:leading-none">
              {name}
            </h2>
            <div className="grid grid-cols-2 gap-1">
              <div className="flex justify-center gap-2 bg-slate-200/40 py-1 dark:bg-slate-600/60">
                <h3 className="text-xs font-normal">Followers</h3>
                <p className="text-xs font-medium">
                  {abbreviateNumber(followers)}
                </p>
              </div>
              <div className="flex justify-center gap-2 bg-slate-200/40 py-1 dark:bg-slate-600/60">
                <h3 className="text-xs font-normal">Popularity</h3>
                <p className="text-xs font-medium">{pop}</p>
              </div>
            </div>
            {genres.length > 0 && (
              <GenreTags
                id={id}
                items={genres}
                className="bg-slate-200/40 text-xs dark:bg-slate-500/80"
              />
            )}

            {/* Spotify Link to Artist */}
            <a
              href={url}
              target="_blank"
              rel="noopener noreferrer"
              className="horizontal-mask flex w-full items-center justify-center gap-0.5 rounded-br-md bg-[#1DB954] text-center text-xs text-black">
              Spotify
              <ExternalLink className="mb-1 h-[.7rem] w-auto" />
            </a>
          </div>
        </div>
        <Image
          className="absolute left-0 top-0 -z-10 h-full w-full rounded-md opacity-80 dark:opacity-40"
          width={100}
          height={100}
          src={img ? img : NO_IMAGE}
          alt={alt}
        />
      </section>
    );
  }
);

ArtistCardHorizontal.displayName = "ArtistCardHorizontal";

export default ArtistCardHorizontal;
