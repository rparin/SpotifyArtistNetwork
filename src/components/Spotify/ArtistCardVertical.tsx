import React from "react";
import Image from "next/image";
import Link from "next/link";
import GenreTags from "./GenreTags";
import abbreviateNumber from "@/utils/abbreviateNumber";
import { ExternalLink } from "lucide-react";
import { NO_IMAGE } from "@/constants/SpotifyConstants";
import { SpotifyArtistCard } from "@/dto/Spotify/SpotifyArtistCardDto";

const ArtistCardVertical = React.forwardRef<HTMLElement, SpotifyArtistCard>(
  ({ id, name, img, genres, followers, pop, url, alt }, ref) => {
    const netLink = `/network/${name}_${id}`;
    return (
      <section
        ref={ref}
        className="relative w-full max-w-[13rem] rounded-md hover:mb-0 hover:bg-blue-950/20 dark:hover:bg-slate-400/20 lg:max-w-[15rem]">
        {/* Artist Music Info */}
        <article className="flex flex-col gap-2 rounded-t-md py-3 backdrop-blur-md">
          <Link
            href={netLink}
            className="flex flex-col items-center text-center">
            <Image
              className="h-28 w-28 rounded-[50%] object-cover lg:h-32 lg:w-32"
              width={100}
              height={100}
              src={img ? img : NO_IMAGE}
              alt={alt}
            />
            <h2 className="horizontal-mask mt-2 line-clamp-1 w-40 bg-slate-200/40 text-2xl font-medium dark:bg-slate-400/20 lg:text-3xl">
              {name}
            </h2>
          </Link>
          <GenreTags
            id={id}
            items={genres}
            className="bg-slate-200/40 dark:bg-slate-500/80"
          />
        </article>

        {/* Artist Social Info */}
        <article>
          <Link
            href={netLink}
            className="grid grid-cols-2 gap-1 backdrop-blur-md">
            <div className="flex flex-col items-center justify-center bg-slate-200/40 py-2 text-center dark:bg-slate-600/60">
              <h3 className="text-sm font-normal lg:text-base">Followers</h3>
              <p className="text-sm font-medium lg:text-base">
                {abbreviateNumber(followers)}
              </p>
            </div>
            <div className="flex flex-col items-center justify-center bg-slate-200/40 py-2 text-center dark:bg-slate-600/60">
              <h3 className="text-sm font-normal lg:text-base">Popularity</h3>
              <p className="text-sm font-medium lg:text-base">{pop}</p>
            </div>
          </Link>
        </article>

        {/* Spotify Link to Artist */}
        <article className="rounded-b-md bg-[#1DB954] text-black">
          <Link
            href={url}
            target="_blank"
            rel="noopener noreferrer"
            className="flex items-center justify-center gap-1 text-center">
            Spotify
            <ExternalLink className="mb-2 h-[.7rem] w-auto" />
          </Link>
        </article>

        <Image
          className="absolute left-0 top-0 -z-10 h-full w-full rounded-md opacity-80 dark:opacity-40"
          width={100}
          height={100}
          src={img ? img : NO_IMAGE}
          alt={alt}
        />
      </section>
    );
  }
);

ArtistCardVertical.displayName = "ArtistCardVertical";

export default ArtistCardVertical;
