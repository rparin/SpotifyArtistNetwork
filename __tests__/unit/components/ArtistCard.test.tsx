import { act, render } from "@testing-library/react";
import { axe, toHaveNoViolations } from "jest-axe";
import ArtistCardVertical from "@/components/Spotify/ArtistCardVertical";
import ArtistCardHorizontal from "@/components/Spotify/ArtistCardHorizontal";
import { SpotifyArtistCard } from "@/dto/Spotify/SpotifyArtistCardDto";

const PROPS: SpotifyArtistCard = {
  url: "https://open.spotify.com/artist/0bAsR2unSRpn6BQPEnNlZm",
  id: "0bAsR2unSRpn6BQPEnNlZm",
  name: "Aimer",
  img: "https://i.scdn.co/image/ab6761610000e5eb23241889efb57a4ce8338932",
  followers: 100,
  pop: 67,
  genres: ["pop", "rock"],
  alt: "Aimer Spotify",
};

describe("testing Artist Card Snapshot", () => {
  test("it, Vertical Card, renders component unchanged", () => {
    const { container } = render(<ArtistCardVertical {...PROPS} />);
    expect(container).toMatchSnapshot();
  });

  test("it, Horizontal Card, renders component unchanged", () => {
    const { container } = render(<ArtistCardHorizontal {...PROPS} />);
    expect(container).toMatchSnapshot();
  });
});

describe("testing Artist Card Accessibility", () => {
  expect.extend(toHaveNoViolations);

  test("it, Vertical Card, should have no accessibility violations", async () => {
    const { container } = render(<ArtistCardVertical {...PROPS} />);
    const results = await act(async () => axe(container));
    expect(results).toHaveNoViolations();
  });

  test("it, Horizontal Card, should have no accessibility violations", async () => {
    const { container } = render(<ArtistCardHorizontal {...PROPS} />);
    const results = await act(async () => axe(container));
    expect(results).toHaveNoViolations();
  });
});
