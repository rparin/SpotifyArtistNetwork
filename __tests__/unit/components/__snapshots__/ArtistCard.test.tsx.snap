// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Artist Card Snapshot it, Horizontal Card, renders component unchanged 1`] = `
<div>
  <section
    class="relative w-[22rem] rounded-md md:w-96"
  >
    <div
      class="flex items-center gap-2 rounded-md px-2 py-2 backdrop-blur-md"
    >
      <img
        alt="Aimer Spotify"
        class="h-20 w-20 rounded-[50%] object-cover md:h-24 md:w-24"
        data-nimg="1"
        decoding="async"
        height="100"
        loading="lazy"
        src="/_next/image?url=https%3A%2F%2Fi.scdn.co%2Fimage%2Fab6761610000e5eb23241889efb57a4ce8338932&w=256&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fi.scdn.co%2Fimage%2Fab6761610000e5eb23241889efb57a4ce8338932&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fi.scdn.co%2Fimage%2Fab6761610000e5eb23241889efb57a4ce8338932&w=256&q=75 2x"
        style="color: transparent;"
        width="100"
      />
      <div
        class="flex w-60 flex-col gap-1 md:w-64"
      >
        <h2
          class="line-clamp-1 flex justify-center text-center text-sm font-medium leading-tight md:text-base md:leading-none"
        >
          Aimer
        </h2>
        <div
          class="grid grid-cols-2 gap-1"
        >
          <div
            class="flex justify-center gap-2 bg-slate-200/40 py-1 dark:bg-slate-600/60"
          >
            <h3
              class="text-xs font-normal"
            >
              Followers
            </h3>
            <p
              class="text-xs font-medium"
            >
              100
            </p>
          </div>
          <div
            class="flex justify-center gap-2 bg-slate-200/40 py-1 dark:bg-slate-600/60"
          >
            <h3
              class="text-xs font-normal"
            >
              Popularity
            </h3>
            <p
              class="text-xs font-medium"
            >
              67
            </p>
          </div>
        </div>
        <div
          class="no-scrollbar group relative w-full overflow-y-hidden px-4"
        >
          <button
            aria-label="Genre Tags: Scroll Left"
            class="invisible absolute left-[-0.12rem] top-[-2px] group-focus-within:visible group-hover:visible"
          />
          <button
            aria-label="Genre Tags: Scroll Right"
            class="invisible absolute right-[-0.12rem] top-[-2px] group-focus-within:visible group-hover:visible"
          >
            <svg
              class="lucide lucide-chevron-right"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m9 18 6-6-6-6"
              />
            </svg>
          </button>
          <ul
            class="no-scrollbar horizontal-mask flex gap-2 overflow-x-scroll whitespace-nowrap px-3 pl-7 text-sm text-black dark:text-white "
            tabindex="0"
          >
            <li
              class="rounded-xl px-3 bg-slate-200/40 text-xs dark:bg-slate-500/80"
            >
              pop
            </li>
            <li
              class="rounded-xl px-3 bg-slate-200/40 text-xs dark:bg-slate-500/80"
            >
              rock
            </li>
          </ul>
        </div>
        <a
          class="horizontal-mask flex w-full items-center justify-center gap-0.5 rounded-br-md bg-[#1DB954] text-center text-xs text-black"
          href="https://open.spotify.com/artist/0bAsR2unSRpn6BQPEnNlZm"
          rel="noopener noreferrer"
          target="_blank"
        >
          Spotify
          <svg
            class="lucide lucide-external-link mb-1 h-[.7rem] w-auto"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15 3h6v6"
            />
            <path
              d="M10 14 21 3"
            />
            <path
              d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
            />
          </svg>
        </a>
      </div>
    </div>
    <img
      alt="Aimer Spotify"
      class="absolute left-0 top-0 -z-10 h-full w-full rounded-md opacity-80 dark:opacity-40"
      data-nimg="1"
      decoding="async"
      height="100"
      loading="lazy"
      src="/_next/image?url=https%3A%2F%2Fi.scdn.co%2Fimage%2Fab6761610000e5eb23241889efb57a4ce8338932&w=256&q=75"
      srcset="/_next/image?url=https%3A%2F%2Fi.scdn.co%2Fimage%2Fab6761610000e5eb23241889efb57a4ce8338932&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fi.scdn.co%2Fimage%2Fab6761610000e5eb23241889efb57a4ce8338932&w=256&q=75 2x"
      style="color: transparent;"
      width="100"
    />
  </section>
</div>
`;

exports[`testing Artist Card Snapshot it, Vertical Card, renders component unchanged 1`] = `
<div>
  <section
    class="relative w-full max-w-[13rem] rounded-md hover:mb-0 hover:bg-blue-950/20 dark:hover:bg-slate-400/20 lg:max-w-[15rem]"
  >
    <header
      class="flex flex-col gap-2 rounded-t-md py-3 backdrop-blur-md"
    >
      <a
        class="flex flex-col items-center text-center"
        href="/network/Aimer_0bAsR2unSRpn6BQPEnNlZm"
      >
        <img
          alt="Aimer Spotify"
          class="h-28 w-28 rounded-[50%] object-cover lg:h-32 lg:w-32"
          data-nimg="1"
          decoding="async"
          height="100"
          loading="lazy"
          src="/_next/image?url=https%3A%2F%2Fi.scdn.co%2Fimage%2Fab6761610000e5eb23241889efb57a4ce8338932&w=256&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fi.scdn.co%2Fimage%2Fab6761610000e5eb23241889efb57a4ce8338932&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fi.scdn.co%2Fimage%2Fab6761610000e5eb23241889efb57a4ce8338932&w=256&q=75 2x"
          style="color: transparent;"
          width="100"
        />
        <h2
          class="horizontal-mask mt-2 line-clamp-1 w-40 bg-slate-200/40 text-xl font-medium dark:bg-slate-400/20"
        >
          Aimer
        </h2>
      </a>
      <div
        class="no-scrollbar group relative w-full overflow-y-hidden px-4"
      >
        <button
          aria-label="Genre Tags: Scroll Left"
          class="invisible absolute left-[-0.12rem] top-[-2px] group-focus-within:visible group-hover:visible"
        />
        <button
          aria-label="Genre Tags: Scroll Right"
          class="invisible absolute right-[-0.12rem] top-[-2px] group-focus-within:visible group-hover:visible"
        >
          <svg
            class="lucide lucide-chevron-right"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m9 18 6-6-6-6"
            />
          </svg>
        </button>
        <ul
          class="no-scrollbar horizontal-mask flex gap-2 overflow-x-scroll whitespace-nowrap px-3 pl-7 text-sm text-black dark:text-white "
          tabindex="0"
        >
          <li
            class="rounded-xl px-3 bg-slate-200/40 backdrop-blur-md dark:bg-slate-500/80"
          >
            pop
          </li>
          <li
            class="rounded-xl px-3 bg-slate-200/40 backdrop-blur-md dark:bg-slate-500/80"
          >
            rock
          </li>
        </ul>
      </div>
    </header>
    <a
      class="grid grid-cols-2 gap-1 backdrop-blur-md"
      href="/network/Aimer_0bAsR2unSRpn6BQPEnNlZm"
      tabindex="-1"
    >
      <div
        class="flex flex-col items-center justify-center bg-slate-200/40 py-2 text-center dark:bg-slate-600/60"
      >
        <h3
          class="text-sm font-normal"
        >
          Followers
        </h3>
        <p
          class="text-sm font-medium"
        >
          100
        </p>
      </div>
      <div
        class="flex flex-col items-center justify-center bg-slate-200/40 py-2 text-center dark:bg-slate-600/60"
      >
        <h3
          class="text-sm font-normal"
        >
          Popularity
        </h3>
        <p
          class="text-sm font-medium"
        >
          67
        </p>
      </div>
    </a>
    <footer
      class="rounded-b-md bg-[#1DB954] text-black"
    >
      <a
        class="flex items-center justify-center gap-1 text-center"
        href="https://open.spotify.com/artist/0bAsR2unSRpn6BQPEnNlZm"
        rel="noopener noreferrer"
        target="_blank"
      >
        Spotify
        <svg
          class="lucide lucide-external-link mb-2 h-[.7rem] w-auto"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15 3h6v6"
          />
          <path
            d="M10 14 21 3"
          />
          <path
            d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
          />
        </svg>
      </a>
    </footer>
    <img
      alt="Aimer Spotify"
      class="absolute left-0 top-0 -z-10 h-full w-full rounded-md opacity-80 dark:opacity-40"
      data-nimg="1"
      decoding="async"
      height="100"
      loading="lazy"
      src="/_next/image?url=https%3A%2F%2Fi.scdn.co%2Fimage%2Fab6761610000e5eb23241889efb57a4ce8338932&w=256&q=75"
      srcset="/_next/image?url=https%3A%2F%2Fi.scdn.co%2Fimage%2Fab6761610000e5eb23241889efb57a4ce8338932&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fi.scdn.co%2Fimage%2Fab6761610000e5eb23241889efb57a4ce8338932&w=256&q=75 2x"
      style="color: transparent;"
      width="100"
    />
  </section>
</div>
`;
